mono/metadata/marshal-ilgen.c

```diff
@@ -1171,7 +1171,15 @@ emit_thread_interrupt_checkpoint_call (MonoMethodBuilder *mb, gpointer checkpoin
 	/* Throw the exception returned by the checkpoint function, if any */
 	mono_mb_emit_byte (mb, CEE_DUP);
 	pos_noex = mono_mb_emit_branch (mb, CEE_BRFALSE);
-	mono_mb_emit_byte (mb, CEE_THROW);
+
+	// This codepath seems to be very rarely hit when using the non-LLVM
+	// backend. I think that the aot compiler passes on the BCL for the default
+	// profiles probably capture this a bit. 
+	//
+	// See calls to mono_get_rethrow_exception_addr in /mini to see the more
+	// common path
+	mono_mb_emit_icall (mb, mono_marshal_rethrow_exception);
+
```

Comment:
why can't we just `mono_mb_emit_byte (mb, CEE_RETHROW);` instead?

Reply:
The exception isn't in flight when we're here. Any attempt to hack it would need to ensure that codegen does not require ECMA behavior, but that it can identify when we're in this case. We need a rethrow-like behavior, while putting a brand new exception into the "current thrown exception" state.

Reply:
I might missing something here: What does "in flight" mean? 

I don't understand the difference: Calling `mono_marshal_rethrow_exception ()` ends up the same as when `OP_RETHROW` is lowered. See AMD64 for example: https://github.com/mono/mono/blob/1d8954083f7da0d83be04abe201a3e833eb8021f/mono/mini/mini-amd64.c#L5010-L5016

Reply:
@lewurm I didn't understand either, but I went and looked at 
https://github.com/mono/mono/blob/8fe989cb70a54d4c46f27de2229895fef5db5635/mono/mini/method-to-ir.c#L11712-L11732

The IL `rethrow` opcode (`CEE_RETHROW`) doesn't take an argument.  it rethrows "the current exception" (because it's supposed to only be used in a catch clause).  By contrast the IR rethrow opcode (`OP_RETHROW`) takes an exception as an operand.

Since marshal-ilgen is writing IL, we can't just use `CEE_RETHROW`.

Reply:
ah, that makes sense, thanks @lambdageek. I think we could instead introduce for example `CEE_MONO_RETHROW` (as far as I understand now, that's what @alexanderkyte meant in his previous comment). But at least an explanatory comment in the source code for that would be nice 🙂